cpp/scenario/gazebo/include/scenario/gazebo/Model.h

```diff
@@ -0,0 +1,222 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SCENARIO_GAZEBO_MODEL_H
+#define SCENARIO_GAZEBO_MODEL_H
+
+#include <ignition/gazebo/Entity.hh>
+#include <ignition/gazebo/EntityComponentManager.hh>
+#include <ignition/gazebo/EventManager.hh>
+
+#include <array>
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+namespace scenario {
+    namespace base {
+        struct ContactData;
+        enum class JointControlMode;
+    } // namespace base
+    namespace gazebo {
+        class Link;
+        class Joint;
+        class Model;
+        class World;
+        using LinkPtr = std::shared_ptr<Link>;
+        using JointPtr = std::shared_ptr<Joint>;
+        using ModelPtr = std::shared_ptr<Model>;
+    } // namespace gazebo
+} // namespace scenario
+
+class scenario::gazebo::Model
+{
+public:
+    Model();
+    virtual ~Model();
+
+    // ============
+    // Gazebo Model
+    // ============
+
+    uint64_t id() const;
+    bool initialize(const ignition::gazebo::Entity modelEntity,
+                    ignition::gazebo::EntityComponentManager* ecm,
+                    ignition::gazebo::EventManager* eventManager);
+    bool createECMResources();
+
+    bool insertModelPlugin(const std::string& libName,
+                           const std::string& className,
+                           const std::string& context = {});
+
+    bool historyOfAppliedJointForcesEnabled() const;
+    bool enableHistoryOfAppliedJointForces( //
+        const bool enable,
+        const size_t maxHistorySizePerJoint = 100);
+    std::vector<double> getHistoryOfAppliedJointForces() const;
+
+    // ==========
+    // Model Core
+    // ==========
+
+    bool valid() const;
+
+    size_t dofs() const;
+    std::string name() const;
+
+    size_t nrOfLinks() const;
+    size_t nrOfJoints() const;
+
+    LinkPtr getLink(const std::string& linkName) const;
+    JointPtr getJoint(const std::string& jointName) const;
+
+    std::vector<std::string> linkNames(const bool scoped = false) const;
+    std::vector<std::string> jointNames(const bool scoped = false) const;
+
+    double controllerPeriod() const;
+    bool setControllerPeriod(const double period);
+
+    // ========
+    // Contacts
+    // ========
+
+    bool contactsEnabled() const;
+    bool enableContacts(const bool enable);
+
+    bool selfCollisions() const;
+    bool enableSelfCollisions(const bool enable);
+
+    std::vector<std::string> linksInContact() const;
+
+    std::vector<base::ContactData>
+    contacts(const std::vector<std::string>& linkNames = {}) const;
+
+    // ==================
+    // Vectorized Methods
+    // ==================
+
+    std::vector<double> jointForces( //
```

Comment:
Same point as https://github.com/robotology/gym-ignition/pull/158/files#r396516195

Reply:
Refer to https://github.com/robotology/gym-ignition/pull/158#discussion_r395493760